### PR TITLE
Memory pair game from LuckyDnepr

### PR DESCRIPTION
**Pull request by LuckyDnepr**
### Memory Pair Game

[Demo](https://luckydnepr.github.io/Memory-game/) |
[Code base](https://github.com/LuckyDnepr/Memory-game)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by LuckyDnepr** [comment]
PR marked as "stale", but there is no review.
Please, review my work.

**Comment by LuckyDnepr** [comment]
UAT done before create PR. Please review.

File: submissions/LuckyDnepr/memory-pair-game/script.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    renderGameField(numberOfCards);
+    generateCardsSet(numberOfCards, themesData[themeIndex].pictures.length - 1);
+    switchMenuAndGameField();
+}
+
+function renderGameField(numberOfCards) {
+    const gameField = doc.querySelector(".game_field");
+    switch (numberOfCards) {
+        case 6:
+            gameField.setAttribute("style", "--cols: 3; --rows: 2");
```
Comment (A-Ostrovnyy) [inline submissions/LuckyDnepr/memory-pair-game/script.js:31]: Avoid inline styles, use CSS classes instead.

File: submissions/LuckyDnepr/memory-pair-game/script.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    if (!card.classList.contains("flipped")) {
+        card.classList.add("flipped");
+        openedCards.push(card);
+        changeCardFace(card);
+        checkPair();
+    }
+}
+
+function checkPair() {
+    if (openedCards.length % 2 == 0) {
```
Comment (A-Ostrovnyy) [inline submissions/LuckyDnepr/memory-pair-game/script.js:60]:
```suggestion
    if (openedCards.length === 2) {
```
Your array.length never will be greater than 2

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
Good work!
Please add files with markup, data, and styles.

**Comment by LuckyDnepr** [comment]
### Thanks for review!
When I look at code from months ago, I want to rewrite *everything* with **new methods**, **new knowledge**.
#### I think this is okay.
## Result:
* I have implemented all the changes you specified. 
* I also found a few mistakes of my own due to inexperience.

File: submissions/LuckyDnepr/memory-pair-game/style.css [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+body {
+    background-color: #ffeee4;
+}
+
+.fader {
+    height: 100vh;
+    width: 100vw;
+    position: absolute;
+    background-color: rgba(0, 0, 0, 0.733);
+    z-index: 1000;
```
Comment (A-Ostrovnyy) [inline submissions/LuckyDnepr/memory-pair-game/style.css:21]: Avoid such a big numbers for `z-index`

File: submissions/LuckyDnepr/memory-pair-game/style.css [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+    align-items: center;
+}
+
+.nav_menu input[type="radio"] {
+    appearance: none;
+    position: fixed;
+    width: 0;
+}
+
+.nav_menu label {
```
Comment (A-Ostrovnyy) [inline submissions/LuckyDnepr/memory-pair-game/style.css:92]: Avoid styling by HTML tags except you add base styles. [Reason](https://frontstuff.io/you-need-to-stop-targeting-tags-in-css)

File: submissions/LuckyDnepr/memory-pair-game/script.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+function resetCardsStage() {
+    document.querySelectorAll(".game_card").forEach((card) => {
+        card.classList.remove("flipped");
+        card.classList.remove("pair");
+    });
+}
+
+function generateCardsSet(sizeOfSet, highestNumber) {
```
Comment (A-Ostrovnyy) [inline submissions/LuckyDnepr/memory-pair-game/script.js:111]:
```suggestion
function generateCardsSet(sizeOfSet, count) {
```
```suggestion
function generateCardsSet(sizeOfSet, quantity) {
```
```suggestion
function generateCardsSet(sizeOfSet, size) {
```

File: submissions/LuckyDnepr/memory-pair-game/script.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    let mixedSetOfPairs = [],
+        setOfPairs = [];
+    for (let i = 0; i < sizeOfSet / 2; i++) {
+        const randomNumber = Math.floor(Math.random() * highestNumber);
+        if (!setOfPairs.includes(randomNumber)) {
+            setOfPairs.push(randomNumber, randomNumber);
+        } else {
+            i--;
+        }
+    }
```
Comment (A-Ostrovnyy) [inline submissions/LuckyDnepr/memory-pair-game/script.js:121]: As for me you should take a copy of the array with images shuffle it and take first `sizeOfSet / 2` elements and then work with this copy. What do you think about that?

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
Good progress!

File: submissions/LuckyDnepr/memory-pair-game/script.js (hunk 6 above)
Comment (LuckyDnepr) [reply submissions/LuckyDnepr/memory-pair-game/script.js:121]: #### Perhaps.
If you take the first **sizeOfSet/2** elements, and then work with them.
You have to somehow get an array of **sizeOfSet**, in which the indexes of two cards are randomly distributed throughout the **sizeOfSet**. Again we'll have to shuffle everything.
If we simply glue the two shuffled **sizeOfSet/2** element arrays together, we get **card[i] === card[sizeOfSet/2 + i]**, which is too simple gameplay-wise.
I spent a lot of time thinking about randomizing the set of cards, and came to the conclusion that there's no right way to do it. You can generate a jumbled array of **2n** cards in different ways.

File: submissions/LuckyDnepr/memory-pair-game/style.css (hunk 4 above)
Comment (LuckyDnepr) [reply submissions/LuckyDnepr/memory-pair-game/style.css]: Why did I use tag-based styling if I added classes to all these elements? The question is more for me. Practice. More practice.

**Pull request closed without merging**